*24 July 2025*
## Filling out the table in answers.md
### Which file's which?
Taking a look at the first reads of each of the files to start with:
R1:
```bash
zcat 1294_S1_L008_R1_001.fastq.gz | head -4
```
output:
```
@K00337:83:HJKJNBBXX:8:1101:1265:1191 1:N:0:1
GNCTGGCATTCCCAGAGACATCAGTACCCAGTTGGTTCAGACAGTTCCTCTATTGGTTGACAAGGTCTTCATTTCTAGTGATATCAACACGGTGTCTACAA
+
A#A-<FJJJ<JJJJJJJJJJJJJJJJJFJJJJFFJJFJJJAJJJJ-AJJJJJJJFFJJJJJJFFA-7<AJJJFFAJJJJJF<F--JJJJJJF-A-F7JJJJ
```

R2:
```bash
zcat 1294_S1_L008_R2_001.fastq.gz | head -4
```
output:
```
@K00337:83:HJKJNBBXX:8:1101:1265:1191 2:N:0:1
NCTTCGAC
+
#AA<FJJJ
```

R3:
```bash
zcat 1294_S1_L008_R3_001.fastq.gz | head -4
```
output:
```
@K00337:83:HJKJNBBXX:8:1101:1265:1191 3:N:0:1
NTCGAAGA
+
#AAAAJJF
```

R4:
```bash
zcat 1294_S1_L008_R4_001.fastq.gz | head -4
```
output:
```
@K00337:83:HJKJNBBXX:8:1101:1265:1191 4:N:0:1
NTTTTGATTTACCTTTCAGCCAATGAGAAGGCCGTTCATGCAGACTTTTTTAATGATTTTGAAGACCTTTTTGATGATGATGATGTCCAGTGAGGCCTCCC
+
#AAFAFJJ-----F---7-<FA-F<AFFA-JJJ77<FJFJFJJJJJJJJJJAFJFFAJJJJJJJJFJF7-AFFJJ7F7JFJJFJ7FFF--A<A7<-A-7--
```

Looks like the files are as such:
 - R1 - read1
 - R2 - index1
 - R3 - index2
 - R4 - read2
which tracks, since that's what we were told was the order in our lectures yesterday.

### Read lengths
now the lengths of each of the reads:
R1:
```bash
zcat 1294_S1_L008_R1_001.fastq.gz | head -2 | tail -1 | wc
```
output:
```
      1       1     102
```
R2:
```bash
zcat 1294_S1_L008_R2_001.fastq.gz | head -2 | tail -1 | wc
```
output:
```
      1       1       9
```

R3:
```bash
zcat 1294_S1_L008_R3_001.fastq.gz | head -2 | tail -1 | wc
```
output:
```
      1       1       9
```

R4:
```bash
zcat 1294_S1_L008_R4_001.fastq.gz | head -2 | tail -1 | wc
```
```
      1       1     102
```
Note that for all of these that there is also a new line character which means that each of these is actually one shorter than the number which was output.

So.... this means that our biological reads are 101 bp long and our index reads are 8 bp long

### Checking PHRED encoding:
```bash
zcat 1294_S1_L008_R1_001.fastq.gz | head -10000 | awk 'NR%4==0' | grep -E '.*#.*' | wc -l
```
output:
```
1600
```
1600 of the quality scores in our first 10000 lines have the '#' symbol, which isn't in Phred+64, so this is Phred+33. Or our data is really messed up, but let's hope that's not the case.

Let's check the other files...

R2:
```bash
zcat 1294_S1_L008_R2_001.fastq.gz | head -10000 | awk 'NR%4==0' | grep -E '.*#.*' | wc -l
```
output:
```
2500
```

R3:
```bash
zcat 1294_S1_L008_R3_001.fastq.gz | head -10000 | awk 'NR%4==0' | grep -E '.*#.*' | wc -l
```
output:
```
2500
```

R4:
```bash
zcat 1294_S1_L008_R4_001.fastq.gz | head -10000 | awk 'NR%4==0' | grep -E '.*#.*' | wc -l
```
output:
```
2500
```

Looks like every reading in the rest of the files had the '#' character, so we can assume they are all encoded in Phred+33.

## Getting the distributions
Made `distributions.md`, and initialized this mamba environment:
```bash
mamba create -n 'demultiplex'
```

installed gzip
```bash
mambas install gzip
```

and making the unit test files (from the 2017_sequencing directory)
```bash
zcat 1294_S1_L008_R1_001.fastq.gz | head -12 > /projects/bgmp/sbergan/bioinfo/Bi622/Demultiplex/test_R1.fq

zcat 1294_S1_L008_R2_001.fastq.gz | head -12 > /projects/bgmp/sbergan/bioinfo/Bi622/Demultiplex/test_R2.fq

zcat 1294_S1_L008_R3_001.fastq.gz | head -12 > /projects/bgmp/sbergan/bioinfo/Bi622/Demultiplex/test_R3.fq

zcat 1294_S1_L008_R4_001.fastq.gz | head -12 > /projects/bgmp/sbergan/bioinfo/Bi622/Demultiplex/test_R4.fq
```
modified the indices of each of these so there will be one unknown, one index jumped, and one correct for both the forward and the reverse compliments

then gzipped them all (from the demultiplex directory):
```bash
gzip test_*
```

got lengths of each of the files as well:
```bash
zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R1_001.fastq.gz | wc -l
```
output:
```
1452986940
```
this is the same value for each of the other fastq files, meaning we have 363246735 reads. Will use this for the python script.

using each of these divided by 4 to make it easier to sum up averages


The script is written -- initially running interactive with a library tqdm which shows iterations through a loop with very low overhead (+60ns/iteration). Hit 250 million iterations in the first file at ~12 minutes, which means this will take a few hours to run.

Will run with sbatch instead of srun... running with script `plot_scores.sh`.


Job `36637905` was ultimately successful -- messed up when renaming files `r0.png` `r1.png`, `r2.png`, and `r3.png` to `read1.png`, `read2.png`, `read3.png`, and `read4.png`.

*30 July 2025*
Wrote `hamming.py` in order to justify having no q score threshold for the indices by plotting hamming distance across all index pair combinations.

*1 Aug 2025*
Ran
```bash
zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R3_001.fastq.gz | sed -n '2~4p' | grep 'N' | wc -l
```
on talapas to answer question 3.

Output:
```
3328051
```

changed `plot_qscores.py` to include read # in title.

Reran.. job # `36736091`.

Submitted/pushed part 1.